lib/jsdom/living/mutation-observer/MutationObserver-impl.js

```diff
@@ -0,0 +1,88 @@
+"use strict";
+
+const { wrapperForImpl } = require("../generated/utils");
+const { mutationObserverList } = require("../helpers/mutation-observers");
+
+// https://dom.spec.whatwg.org/#mutationobserver
+class MutationObserverImpl {
+  // https://dom.spec.whatwg.org/#dom-mutationobserver-mutationobserver
+  constructor(args) {
+    const [callback] = args;
+
+    this._callback = callback;
+    this._nodeList = [];
+    this._recordQueue = [];
+
+    mutationObserverList.push(this);
```

Comment:
The fact that this never gets deleted is a bit worrying. A program that adds lots of MOs will seem to "leak" memory, even if it loses all references to them. I filed https://github.com/whatwg/dom/issues/720 to check.

I don't think there's any solution for this except either (a) waiting for weak references to be added to JS; (b) implementing a maximum number of them, like we did with concurrentNodeIterators (no longer a publicly tune-able option).

I'm not sure what we should do exactly, but probably we should at least add a note to the README. I'm asking people on Twitter what they think... https://mobile.twitter.com/domenic/status/1069393241131401217

Reply:
I get around this issue by avoiding having to keep a list of all the MOs. At the end of the day, the list is only used to keep the order of the mutation observer.

Instead, now each new MO get assigned a unique incrementing id, and we keep a set of MO that needs to be invoked. We can then invoke the MO in the order based on their id. You can find the details in: https://github.com/jsdom/jsdom/pull/2398/commits/02340384f5d680b4f4e99084d009ab014058b3d3